homeassistant/components/deluge/__init__.py

```diff
@@ -1 +1,356 @@
+)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Import the Deluge component from config."""
+    if DOMAIN in config:
+        for entry in config[DOMAIN]:
+            hass.async_create_task(
+                hass.config_entries.flow.async_init(
+                    DOMAIN, context={"source": SOURCE_IMPORT}, data=entry
+                )
+            )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Deluge from a config entry."""
+    client = DelugeClient(hass, entry)
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = client
+
+    if not await client.async_setup():
+        return False
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    for platform in PLATFORMS:
+        await hass.config_entries.async_forward_entry_unload(entry, platform)
+
+    client = hass.data[DOMAIN].pop(entry.entry_id)
+    if client.unsub_timer:
+        client.unsub_timer()
+
+    return True
+
+
+async def get_api(hass, entry):
+    """Get Transmission client."""
+    host = entry[CONF_HOST]
+    port = entry[CONF_PORT]
+    username = entry.get(CONF_USERNAME)
+    password = entry.get(CONF_PASSWORD)
+
+    try:
+        api = await hass.async_add_executor_job(
+            DelugeRPCClient, host, port, username, password, True
+        )
+        api.connect()
+        _LOGGER.debug("Successfully connected to %s", host)
+        return api
+
+    except ConnectionRefusedError as error:
+        _LOGGER.error(error)
+        raise CannotConnect
+    except Exception as error:  # pylint: disable=broad-except
+        _LOGGER.error(error)
+        if "Username does not exist" in str(error):
+            raise UserNameError
+        if "Password does not match" in str(error):
+            raise PasswordError
+        raise UnknownError
+
+
+class DelugeClient:
+    """Deluge Client Object."""
+
+    def __init__(self, hass, config_entry):
+        """Initialize the Deluge client."""
+        self.hass = hass
+        self.config_entry = config_entry
+        self._deluge_data = None
+        self.unsub_timer = None
+
+    @property
+    def api(self):
+        """Return the _deluge_data object."""
+        return self._deluge_data
+
+    async def async_setup(self):
+        """Set up the Deluge client."""
+
+        try:
+            deluge_api = await get_api(self.hass, self.config_entry.data)
+        except CannotConnect:
+            raise ConfigEntryNotReady
+        except (UserNameError, PasswordError, UnknownError):
+            return False
+
+        self._deluge_data = DelugeData(self.hass, self.config_entry, deluge_api)
+
+        await self.hass.async_add_executor_job(self._deluge_data.init_torrent_list)
+        await self.hass.async_add_executor_job(self._deluge_data.update)
+        self.add_options()
+        self.set_scan_interval(self.config_entry.options[CONF_SCAN_INTERVAL])
```

Comment:
Let's start with making scan interval not configurable. Generally we know the right amount as a developer and don't want inexperienced users allow to overburden the server.

Reply:
I recommend that we still give the user this option as sometimes they need a faster update interval when monitoring torrents.

Reply:
scan interval is not configurable during setup. it is only available as an option after the config entry is successful.